### ISSUE TITLE
After Backup-restore group messages are send to unintended recipients

### ISSUE THREAD

**Issue by fsck** [signalapp/Signal-Android#7909]
- [X] I have searched open and closed issues for duplicates
- [X] I am submitting a bug report for existing functionality that does not work as intended
- [X] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [X] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
After restoring a Backup in Signal I sent a delighted message to a group of fellow signal-enthusiasts. To my dismay I had to discover that this message was not only delivered to these fine individuals, but also to an entirely different group which I converse with. The overlap of these two groups is included twice in the delivery receipts. This caused an "Unnamed Group" and the message in question to appear for the users which are not in the intended group.

After I sent a group update from the faulty signal device I again am able to send messages to the correct group of people.

Additionally now we have both groups messed up and receiving an alarming number of disappearing messages timer upgrades from users of the unintended group. This might be a separate issue I will raise with the Signal-Desktop repository.

### Steps to reproduce
- Restore (a rather large >160000Messages, >4GB) Backup
- Send Message to a group

**Actual result:** Message received by multiple groups
**Expected result:** Message received in a single group only

### Screenshots
![screenshot_20180619-160359](https://user-images.githubusercontent.com/17763262/41610409-6bb42fa6-73ed-11e8-8afd-02a8b5565766.png)
The intended group with the offending message and the disappearing message timer from non-members.

![signal-2018-06-18-224938](https://user-images.githubusercontent.com/17763262/41610439-80b3eaea-73ed-11e8-9ede-d05d885d9432.png)
The unnamed group on the phone of one unintended recipient

![screenshot_20180619-002611](https://user-images.githubusercontent.com/17763262/41610676-1de3b912-73ee-11e8-8fd4-1e3d00f3c26f.png)
Message self-identifying as send to a larger than intended group of people

![screenshot_20180619-002633](https://user-images.githubusercontent.com/17763262/41610739-459fc36a-73ee-11e8-9075-0750f43c7224.png)
After a group update with the intended set of people

### Device info
**Device:** Motorola Moto G6+
**Android version:** 8.0.0
**Signal version:** 4.20.8

### Link to debug log
https://debuglogs.org/7c27e3f93e79c6bb41d1bc71feb1225105e941de8dba6f8296f18ddaad2497ac

**Comment by greyson-signal** [comment]
Hi @fsck! Thank you very much for the detailed report! However, I'm going to ask for some clarifications just so I can get the sequence of events straight. The logs are capturing a ton here, unfortunately.

Can you help me understand the following:

1. Your app version where you made the backup.
2. Your group state at the time of backup (it sounds like there's two groups specifically in question here?). Specifically the people in the group, and what the expiration timers were.
3. Your group state right after restore. I don't know if you saw this before you started messaging, but was the list of people in the group correct? i.e if you clicked the group icon on the action bar, was that listing correct? I know when you sent the message it went to more people, I guess I'm just curious if those two listings are out of sync.
4. A walkthrough of the messages you sent to what group, who they went to, who they were _supposed_ to go to, and any side effects (like the creation of the new unnamed group).
5. Whether you all use the desktop client as well, and whether you were using the desktop client in between when the backup was created and when you restored the backup.

It'd be super helpful if you could use names in the form of aliases. Examples would be:

> Group state before backup:
> Group A: Alice, Bob
> Group B: Carol, Dan, Eric
> Expiring message sent to 1 week
> 
> ...
> 
> I sent a message to Group A, but when I viewed the message details, the receiver list was "Alice, Bob, Alice, Bob, Carol, Dan, Eric".
> Afterwards, an "Unnnamed group" was created with just Dan.

Thanks again! Sorry if this is a lot, it's just important to get the details straight when investigating issues like this.

**Comment by fsck** [comment]
Hi @greyson-signal, thanks for the prompt feedback.

1. 4.19.3

2. Group state before backup:
Group A: (F)sck (R)obert* (C)ollin* expiring 1 day
Group B: same as above plus (B)erry* (L)ars* (A)lex* and about 30 other people. Expiring 1 week

3. Group state after backup:
Seems fine. User listing appears nominal. Except Messages in Group A go to Group B as well. After Group update sent by the offending device it is back to the original behavior. So when I send a message in Group A:
Group A[as shown in the status bar]: (F)sck (R)obert* (C)ollin* expiring 1 day
Group A[Receiverlist]: (R)obert* (C)ollin* plus (R)obert^ (C)ollin^ (B)erry* (L)ars* (A)lex* and about 30 other people. Expiring 1 day
An unnamed group was created for everyone except members of the group A.

*: Names changed
^: yes, actually twice in the sent message list

4. See 3. and the corresponding screenshot in the original Issue. Message at 22:31h is the problematic one. The one at 22:34h is after the group update.

5. Yes and probably yes during the backup as it takes multiple hours. But I do not think I have had it running during or after the restore, much less to Relink it with the new phone.

**Comment by greyson-signal** [comment]
Thank you very much for the details! This is very helpful. Some more follow-up questions :)

1. Is this issue reproducible? i.e. If you restore from the backup again, would this same thing happen?
2. After restoring from the backup, did you link to desktop before sending your first problem message? If I'm reading your point (5) correctly, I believe the answer is yes, but just wanted to be sure.
3. What "group update" are you referring to in point (3)? Are you adding people, changing the title, setting the timer, etc?

**Comment by fsck** [comment]
1. I hope you understand that I will definitely not attempt to reproduce this issue or restore any backups until this is fixed. I have been using Signal for a number of years and migrated between phones with my entire database many times (using adb) without any issues. This is the first time that I used the official backup/migration and signal has violated every promise of integrity and confidentiality there was.
2. According to the logs my colleague has reviewed I have linked the desktop before I sent the first message. But just to be clear: I have not been sending messages prior to that. The “first problematic message” was the very first message in a group.
3. See “You updated the group” in the first screenshot. I simply opened the *Edit group* dialog and pressed the check mark in the upper right corner. This is probably a bug as well but it causes a group update without updating anything.
